vespaclient-container-plugin/src/main/java/com/yahoo/document/restapi/DocumentOperationExecutor.java

```diff
@@ -0,0 +1,585 @@
+                Thread.sleep(sleepMillis);
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+
+    /** The executor will call <em>exactly one</em> callback <em>exactly once</em> for contexts submitted to it. */
+    private static class Context<T> {
+
+        private final AtomicBoolean handled = new AtomicBoolean();
+        private final BiConsumer<ErrorType, String> onError;
+        private final Consumer<T> onSuccess;
+
+        Context(BiConsumer<ErrorType, String> onError, Consumer<T> onSuccess) {
+            this.onError = onError;
+            this.onSuccess = onSuccess;
+        }
+
+        void error(ErrorType type, String message) {
+            if ( ! handled.getAndSet(true))
+                onError.accept(type, message);
+        }
+
+        void success(T result) {
+            if ( ! handled.getAndSet(true))
+                onSuccess.accept(result);
+        }
+
+        boolean handled() {
+            return handled.get();
+        }
+
+    }
+
+    /** Context for reacting to the progress of a visitor session. Completion signalled by an optional progress token. */
+    public static class VisitorContext extends Context<Optional<String>> {
+
+        private final Consumer<Document> onDocument;
+
+        public VisitorContext(BiConsumer<ErrorType, String> onError, Consumer<Optional<String>> onSuccess, Consumer<Document> onDocument) {
+            super(onError, onSuccess);
+            this.onDocument = onDocument;
+        }
+
+        void document(Document document) {
+            if ( ! handled())
+                onDocument.accept(document);
+        }
+
+    }
+
+    /** Context for a document operation. */
+    public static class OperationContext extends Context<Optional<Document>> {
+
+        public OperationContext(BiConsumer<ErrorType, String> onError, Consumer<Optional<Document>> onSuccess) {
+            super(onError, onSuccess);
+        }
+
+    }
+
+
+    public enum ErrorType {
+        OVERLOAD,
+        NOT_FOUND,
+        PRECONDITION_FAILED,
+        BAD_REQUEST,
+        TIMEOUT,
+        ERROR;
+    }
+
+
+    private static class Delayed {
+
+        private final Supplier<Result> operation;
+        private final OperationContext context;
+        private final Instant readyAt;
+
+        Delayed(Instant readyAt, Supplier<Result> operation, OperationContext context) {
+            this.readyAt = requireNonNull(readyAt);
+            this.context = requireNonNull(context);
+            this.operation = requireNonNull(operation);
+        }
+
+        Supplier<Result> operation() { return operation; }
+        OperationContext context() { return context; }
+        Instant readyAt() { return readyAt; }
+
+    }
+
+
+    public static class VisitorOptions {
+
+        private final Optional<String> cluster;
```

Comment:
Ah, yes ... I find myself using `Optional` quite liberally, but I'm not averse to doing null checks either. Since all uses of these fields are private, I suppose `null` is really the better choice.

Reply:
I think I'll leave it like this this time, but I'll consider `null` the next time. The code to generate the `VisitorParameters` is indeed simpler with `Optional`s.